### PR TITLE
tools: EvalDelta and ConsensusParams x-repo type checks

### PR DESCRIPTION
**Pull request by tzaffi**
### 1. Introduce recursive `go test ./...`  in CI and `go mod tidy` both in `Makefile` and `CI`
It turns out that the **x-repo-types** tests weren't actually running in C.I. This is because `go test ./...` (and `gotestsum`) don't recurse into module directories (those that have a `go.mod` file). 

We can see that [this commit](https://github.com/algorand/go-algorand/pull/5381/commits/07934cdda923e455f2ed2c57bf459bec32fc5426) very happily passed CI:
image

#### References
* [StackOverflow](https://stackoverflow.com/questions/61163306/running-go-test-in-parent-directory-of-multiple-go-modules)
* [gotestsum](https://github.com/gotestyourself/gotestsum/issues/231)

### Summary of CI/Makefile Changes:
* Introducing `.github/workflows/tools.yml` - (thanks @cce cf. https://github.com/tzaffi/go-algorand/pull/17)
* `Makefile` - Recursive recipe for `make tidy`

### 2. Introduce more cross-type checks
* `EvalDelta` in `data.transactions` vs. in `go-algorand-sdk`'s `types`.
* `ConsensusParams` in `config` vs. in `go-algorand-sdk`'s `protocol.config`

## Context

`EvalDelta` was introduced to fix failing Indexer end-to-end tests. See:
* https://github.com/algorand/indexer/issues/1521
* https://github.com/algorand/go-algorand-sdk/pull/531
* https://github.com/algorand/conduit/pull/73

`ConsensusParams` was moved from indexer to go-algorand-sdk in https://github.com/algorand/go-algorand-sdk/pull/543

## Related Issues

* #5316 
* #5427

### Which tests were being skipped and what is the impact going forward?

* ~cmd/partitiontest_linter/linter_test.go~ - **STILL SKIPPED GOING FORWARD**
* tools/block-generator/generator/generate_test.go
* tools/block-generator/generator/server_test.go
* tools/block-generator/generator/utils_test.go
* tools/x-repo-types/typeAnalyzer/typeAnalyzer_test.go
* tools/x-repo-types/xrt_test.go

## Testing

Purely on CI

### CODE REVIEW

File: tools/x-repo-types/xrt_test.go [hunk 1]
```diff
@@ -87,7 +95,7 @@ func TestCrossRepoTypes(t *testing.T) {
 			yType: "StateProof",
 		},
 	}
-
+	testCases = testCases[3:4]
```
Comment (tzaffi) [inline tools/x-repo-types/xrt_test.go:98]:
```suggestion
```

**Review by winder — DISMISSED** [review]
This seems reasonable.

File: tools/x-repo-types/xrt_test.go [hunk 2]
```diff
@@ -87,7 +95,6 @@ func TestCrossRepoTypes(t *testing.T) {
 			yType: "StateProof",
 		},
 	}
```
Comment (tzaffi) [inline tools/x-repo-types/xrt_test.go:97]:
```suggestion
	}

```

**Review by winder — DISMISSED** [review]

File: Makefile [hunk 3]
```diff
@@ -245,8 +245,8 @@ $(GOPATH1)/bin/%:
 	cp -f $< $@
 
 test: build
-	$(GOTESTCOMMAND) $(GOTAGS) -race $(UNIT_TEST_SOURCES) -timeout 1h -coverprofile=coverage.txt -covermode=atomic
-
+	find . -name 'go.mod' -execdir $(GOTESTCOMMAND) $(GOTAGS) -race $(UNIT_TEST_SOURCES) -timeout 1h -coverprofile=coverage.txt -covermode=atomic \;
+
```
Comment (tzaffi) [inline Makefile:249]:
```suggestion

```

File: tools/x-repo-types/xrt_test.go [hunk 4]
```diff
@@ -63,6 +63,14 @@ func TestCrossRepoTypes(t *testing.T) {
...
 			skipReason: `Several issues. For example: LEVEL 5 of goal bookkeeping.Block is EvalDelta with field [SharedAccts](codec:"sa,allocbound=config.MaxEvalDeltaAccounts") VS SDK types.EvalDelta is missing SharedAccts field`,
 		},
+		{
+			name:    "goal-v-sdk-eval-delta",
+			xPkg:    "github.com/algorand/go-algorand/data/transactions",
+			xType:   "EvalDelta",
+			yPkg:    "github.com/algorand/go-algorand-sdk/v2/types",
+			yBranch: "develop",
+			yType:   "EvalDelta",
+		},
```
Comment (tzaffi) [inline tools/x-repo-types/xrt_test.go:73]: TODO: don't forget `ConsensusParams` in `config` vs. in go-algorand-sdk's `protocol.config`

File: Makefile [hunk 5]
```diff
@@ -80,6 +80,9 @@ endif
 GOTAGS      := --tags "$(GOTAGSLIST)"
 GOTRIMPATH	:= $(shell GOPATH=$(GOPATH) && go help build | grep -q .-trimpath && echo -trimpath)
 
+COVER_PROFILE  := $(SRCPATH)/coverage.txt
+COVER_OUT      := $(SRCPATH)/cover.out
```
Comment (tzaffi) [inline Makefile:84]: Why do we need two different coverage files?

File: Makefile [hunk 6]
```diff
@@ -244,17 +270,21 @@ $(GOPATH1)/bin/ddconfig.sh: scripts/ddconfig.sh
 $(GOPATH1)/bin/%:
 	cp -f $< $@
 
-test: build
-	$(GOTESTCOMMAND) $(GOTAGS) -race $(UNIT_TEST_SOURCES) -timeout 1h -coverprofile=coverage.txt -covermode=atomic
+# run-test on all eligible sub go modules:
+run-test:
+	$(call SUBMODULE_TESTER,$(GOTESTCOMMAND) $(GOTAGS) -race -timeout 1h -coverprofile=coverage.txt -covermode=atomic)
```
Comment (tzaffi) [inline Makefile:275]:
```suggestion
	$(call SUBMODULE_TESTER,$(GOTESTCOMMAND) $(GOTAGS) -race -timeout 1h -coverprofile=$(COVER_PROFILE) -covermode=atomic)
```

File: Makefile (hunk 5 above)
Comment (tzaffi) [reply Makefile:84]: @cce @onetechnical - Any thoughts about this ^^^

File: Makefile [hunk 7]
```diff
@@ -114,20 +131,29 @@ check_go_version:
...
 
 check_shell:
 	find . -type f -name "*.sh" -exec shellcheck {} +
 
 sanity: fix lint fmt tidy
 
+# cover all eligible sub go modules:
 cover:
-	go test $(GOTAGS) -coverprofile=cover.out $(UNIT_TEST_SOURCES)
+	$(call SUBMODULE_TESTER,go test $(GOTAGS) -coverprofile=$(COVER_OUT))
```
[the remark this answers is not in the document]
Comment (tzaffi) [reply Makefile:150]: @cce @onetechnical ^^^

File: tools/x-repo-types/typeAnalyzer/typeAnalyzer.go [hunk 8]
```diff
@@ -32,6 +32,9 @@ import (
 var diffExclusions = map[string]bool{
 	// MicroAlgos is a struct with custom marshal override in go-algorand. In other repos it is a uint64.
 	`github.com/algorand/go-algorand/data/basics :: "basics.MicroAlgos" (struct)`: true,
+
```
Comment (tzaffi) [inline tools/x-repo-types/typeAnalyzer/typeAnalyzer.go:35]: The changes in `x-repo-types` are totally orthogonal to the CICD changes. I'm happy to break up the PR into 2 seperate ones, if that eases the review process.

File: tools/debug/transplanter/main.go [hunk 9]
```diff
@@ -212,18 +212,18 @@ func transcribeSnappyLog(filePath string, output chan txGroupItem, wg *sync.Wait
 	}
 	defer file.Close()
 
-	var headerDecoder headerDecoder
+	var hdrDecoder headerDecoder
```
Comment (tzaffi) [inline tools/debug/transplanter/main.go:215]: Addresses [reviewdog failure](https://github.com/algorand/go-algorand/actions/runs/5156708833/jobs/9288071892?pr=5381#step:5:89)

@algorandskiy

File: .circleci/config.yml [hunk 10]
```diff
@@ -602,11 +602,16 @@ commands:
...
+            GOMOD_DIRS="$(find . -name 'go.mod' -not -path './cmd/partitiontest_linter/*' -exec dirname {} \;)"
+            for gmdir in $GOMOD_DIRS; do
+                pushd $gmdir
+                PACKAGES="$(go list ./... | grep -v /go-algorand/test/)"
+                export PACKAGE_NAMES=$(echo $PACKAGES | tr -d '\n')
+                gotestsum --format standard-verbose --junitfile << parameters.result_path >>/<< parameters.result_subdir >>/${CIRCLE_NODE_INDEX}/results.xml --jsonfile << parameters.result_path >>/<< parameters.result_subdir >>/${CIRCLE_NODE_INDEX}/testresults.json -- --tags "sqlite_unlock_notify sqlite_omit_load_extension" << parameters.short_test_flag >> -race -timeout 1h -coverprofile=coverage.txt -covermode=atomic -p 1 $PACKAGE_NAMES
+                popd
+            done
```
[the remark this answers is not in the document]
Comment (tzaffi) [reply .circleci/config.yml:614]: That sounds like a good proposal. I'll probably revert this test, and introduce a simple pure nested module tester on GA.

File: .github/workflows/tools.yml [hunk 11]
```diff
@@ -0,0 +1,42 @@
+# Test tools directories that have their own go.mod files, and
+# otherwise wouldn't get tested along with the main go-algorand build.
+name: Test tools modules
```
Comment (algorandskiy) [inline .github/workflows/tools.yml:3]: I'm not sure if testing two tools subdirectories need to build entire algorand on three supported platforms, it looks as overkill

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**